### PR TITLE
Captcha Validation for Forms

### PR DESCRIPTION
**Pull request by hannessolo**
Taking a shot at implementing #322 .

### CODE REVIEW

File: src/forms-pipe.js [hunk 1]
```diff
@@ -121,6 +133,22 @@ export async function formsPipe(state, req) {
...
+  const { fetch, config } = state;
+  const { 'captcha-secret-key': captchaSecretKey, 'captcha-type': captchaType } = config;
+  if (captchaType && captchaType !== 'reCaptcha v2') {
+    return error(log, `The captcha type ${captchaType} you have configured is not currently supported.`, 500, res);
+  }
+  if (captchaType && !captchaSecretKey) {
+    return error(log, 'You must configure ', 500, res);
+  }
+  if (captchaType) {
+    const captchaPassed = await verifyCaptcha(fetch, req.headers.get('x-google-captcha-token'), captchaSecretKey);
```
Comment (tripodsan) [inline src/forms-pipe.js:146]: I think the captcha response is included in the `g-recaptcha-response` form field, certainly not the header.

**Review by tripodsan — CHANGES_REQUESTED** [review]
LGTM, but somehow I don't like the return values from `validateCaptcha()`
maybe throw a `PipelineStatusError` instead?

File: src/steps/validate-captcha.js [hunk 2]
```diff
@@ -9,69 +9,58 @@
...
+  }
+
   const response = await fetch('https://www.google.com/recaptcha/api/siteverify', {
     method: 'POST',
-    body: new URLSearchParams({ secret: secretKey, response: token }),
+    body: new URLSearchParams({ secret: secretKey, response: captchaToken }),
   });
   if (!response.ok) {
-    return false;
+    throw new PipelineStatusError(response.status, `Captcha validation request returned ${response.status}`);
```
Comment (tripodsan) [inline src/steps/validate-captcha.js:28]: I don't think that relying the response.status from google is a good idea. either use a `500` or `503`

File: src/steps/validate-captcha.js [hunk 3]
```diff
@@ -9,69 +9,58 @@
...
-    }
-
-    const captchaPassed = await performValidation(fetch, captchaToken, captchaSecretKey);
-    return {
-      success: captchaPassed,
-      message: captchaPassed ? null : 'Captcha validation failed.',
-      status: captchaPassed ? null : 400,
-    };
+  // Handle cases where captcha is not configured correctly
+  if (!Object.keys(SUPPORTED_CAPTCHA_TYPES).includes(captchaType)) {
```
Comment (tripodsan) [inline src/steps/validate-captcha.js]:
```suggestion
  if (!captchaType in SUPPORTED_CAPTCHA_TYPES) {
```

File: src/steps/validate-captcha.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  // Handle cases where captcha is not configured correctly
+  if (!(captcha.type in SUPPORTED_CAPTCHA_TYPES)) {
+    throw new PipelineStatusError(500, `The captcha type ${captcha.type} is not supported.`);
+  }
+  if (!captcha.secret) {
+    throw new PipelineStatusError(500, 'Captcha secret key is not configured.');
+  }
+
+  // Perform validation
+  const validator = SUPPORTED_CAPTCHA_TYPES[captcha.type];
```
Comment (tripodsan) [inline src/steps/validate-captcha.js:64]:
```suggestion
  // Handle cases where captcha is not configured correctly
  const validator = SUPPORTED_CAPTCHA_TYPES[captcha.type];
  if (!validator) {
    throw new PipelineStatusError(500, `The captcha type ${captcha.type} is not supported.`);
  }
  if (!captcha.secret) {
    throw new PipelineStatusError(500, 'Captcha secret key is not configured.');
  }

  // Perform validation
```

**Review by tripodsan — APPROVED** [review]
thanks. LGTM!

File: src/forms-pipe.js [hunk 5]
```diff
@@ -121,6 +122,20 @@ export async function formsPipe(state, req) {
...
+    return error(log, err.message, 400, res);
+  }
+
+  // verify captcha if needed
+  try {
+    await validateCaptcha(state, body);
+  } catch (e) {
+    return error(log, e.message, e.code, res);
+  }
+
```
Comment (tripodsan) [inline src/forms-pipe.js:138]: @hannessolo can you move this after the check below? this avoids extra processing for invalid posts.

Comment (hannessolo) [reply src/forms-pipe.js]: Sure - done.

**Pull request merged**